llpc/patch/llpcPatchDescriptorLoad.cpp

```diff
@@ -136,46 +136,270 @@ bool PatchDescriptorLoad::runOnModule(
+        std::tie(pTopNode, pNode) = m_pPipelineState->FindResourceNode(ResourceMappingNodeType::DescriptorResource,
+                                                                       ResourceMappingNodeType::DescriptorResource,
+                                                                       descSet,
+                                                                       binding);
     }
 
-    assert(pLoadPtr->getCalledFunction()->getName().startswith(LlpcName::DescriptorGetPtrPrefix));
+    Value* pDescPtrAndStride = nullptr;
+    if (pNode == nullptr)
+    {
+        // We did not find the resource node. Use an undef value.
+        pDescPtrAndStride = UndefValue::get(pDescPtrCall->getType());
+    }
+    else
+    {
+        // Get the descriptor pointer and stride as a struct.
+        pDescPtrAndStride = GetDescPtrAndStride(resType, descSet, binding, pTopNode, pNode, shadow, builder);
+    }
+    pDescPtrCall->replaceAllUsesWith(pDescPtrAndStride);
+    m_descLoadCalls.push_back(pDescPtrCall);
+    m_changed = true;
+}
 
-    uint32_t descSet = cast<ConstantInt>(pLoadPtr->getOperand(0))->getZExtValue();
-    uint32_t binding = cast<ConstantInt>(pLoadPtr->getOperand(1))->getZExtValue();
-    Value* pDesc = LoadDescriptor(*pLoadPtr, descSet, binding, pIndex, pLoadFromPtr);
+// =====================================================================================================================
+// Get a struct containing the pointer and byte stride for a descriptor
+Value* PatchDescriptorLoad::GetDescPtrAndStride(
+    ResourceMappingNodeType resType,    // Resource type
+    uint32_t                descSet,    // Descriptor set
+    uint32_t                binding,    // Binding
+    const ResourceNode*     pTopNode,   // Node in top-level descriptor table (TODO: nullptr for shader compilation)
+    const ResourceNode*     pNode,      // The descriptor node itself (TODO: nullptr for shader compilation)
+    bool                    shadow,     // Whether to load from shadow descriptor table
+    IRBuilder<>&            builder)    // [in/out] IRBuilder
+{
+    // Determine the stride if possible without looking at the resource node.
+    uint32_t byteSize = 0;
+    Value* pStride = nullptr;
+    switch (resType)
+    {
+    case ResourceMappingNodeType::DescriptorBuffer:
+    case ResourceMappingNodeType::DescriptorTexelBuffer:
+        byteSize = DescriptorSizeBuffer;
+        if ((pNode != nullptr) && (pNode->type == ResourceMappingNodeType::DescriptorBufferCompact))
+        {
+            byteSize = DescriptorSizeBufferCompact;
+        }
+        pStride = builder.getInt32(byteSize);
+        break;
+    case ResourceMappingNodeType::DescriptorSampler:
+        byteSize = DescriptorSizeSampler;
+        break;
+    case ResourceMappingNodeType::DescriptorResource:
+    case ResourceMappingNodeType::DescriptorFmask:
+        byteSize = DescriptorSizeResource;
+        break;
+    default:
+        llvm_unreachable("");
```

Comment:
-> llvm_unreachable("Should never be called!")
Just like what you have done for existing codes. If this is too verbose, maybe you can change all of them to llvm_unreachable(""). Anyway, I just want to make codes consistent without varying comments for such assert.

Reply:
The only reason all the other ones have the same message is that they replace an LLPC macro that hard-wired that message. Any new one like this should have either no message or (preferably) a meaningful message. See the llvm coding standards doc and codebase. I have now added a meaningful message here.